### ISSUE TITLE
Some Tracelogger statements affect performance despite loglevel debug

### ISSUE THREAD

**Issue by Morlic-fo** [freeorion/freeorion#2604]
### Environment
* **FreeOrion Version**: master as of eb42b0e
* **Operating System**: Win10
* **Fetched as**
  * Compiled from source

### Description
Despite having all log levels set to debug, at least some TraceLogger statements are significantly impacting performance. In particular, there are the statements added by @geoffthemedio in https://github.com/freeorion/freeorion/commit/b7cd0fef4f44652d5520632f5b210d9958be1d7e - not sure if there are any other.

The following profiling results are for a the standard MSVC2017 solution Release build. Setup was starting a game with default settings, enqueue SuperTester building to get full map vision. After this setup, profiler was attached to FreeOrion.exe process and turns auto-advanced. Profiling duration was 15 seconds.
The percentages displayed are relative to total CPU time spent in FreeOrion.exe.
![TraceLogger](https://user-images.githubusercontent.com/11706951/67437941-6eb8b400-f5f2-11e9-85bc-610b9ef81ec8.PNG)

In a next test, I used the following diff to very that the lambda is indeed executed and this is not a profiling error:
```
diff --git a/universe/Condition.cpp b/universe/Condition.cpp
index 6397d6b24..8ba07303c 100644
--- a/universe/Condition.cpp
+++ b/universe/Condition.cpp
@@ -9549,6 +9549,7 @@ void And::Eval(const ScriptingContext& parent_context, ObjectSet& matches,

     auto ObjList = [](const ObjectSet& objs) -> std::string {
         std::stringstream ss;
+        ErrorLogger() << "This should not be called";
         for (const auto& obj : objs)
             ss << obj->Name() << " (" << std::to_string(obj->ID()) << ")  ";
         return ss.str();
```
and sure enough
```
23:24:03.773264 {0x00000764} [error] client : condition.cpp:9552 : This should not be called
```

### Expected Result
Logger statements should be lazily evaluated and not affect performance if the relevant log level is not set.

**Comment by geoffthemedio** [comment]
At least the trivial logger is supposed to be lazy: https://www.boost.org/doc/libs/1_67_0/libs/log/doc/html/log/tutorial/trivial_filtering.html

Possibly in general it doesn't though?: 
https://www.boost.org/doc/libs/1_62_0/libs/log/doc/html/log/rationale/why_not_lazy_streaming.html

Are you sure the inner error logger doesn't override the lazy evaluation filtering somehow?

**Comment by Morlic-fo** [comment]
After some more debugging, seems like a threading-related bug.

```
diff --git a/universe/Condition.cpp b/universe/Condition.cpp
index 6397d6b24..c5bd02528 100644
--- a/universe/Condition.cpp
+++ b/universe/Condition.cpp
@@ -9547,8 +9547,11 @@ void And::Eval(const ScriptingContext& parent_context, ObjectSet& matches,
         }
     }

-    auto ObjList = [](const ObjectSet& objs) -> std::string {
+
+    int counter = 0;
+    auto ObjList = [&counter](const ObjectSet& objs) -> std::string {
         std::stringstream ss;
+        counter++;
         for (const auto& obj : objs)
             ss << obj->Name() << " (" << std::to_string(obj->ID()) << ")  ";
         return ss.str();
@@ -9558,6 +9561,8 @@ void And::Eval(const ScriptingContext& parent_context, ObjectSet& matches,
                             << " with input matches (" << matches.size() << "): " << ObjList(matches)
                             << " and input non_matches(" << non_matches.size() << "): " << ObjList(non_matches);

+    ErrorLogger() << "Post-first: " << counter;
+
     if (search_domain == NON_MATCHES) {
         ObjectSet partly_checked_non_matches;
         partly_checked_non_matches.reserve(non_matches.size());
```


```
15:26:45.950385 {0x000018c4} [error] client : condition.cpp:9566 : Post-first: 0
15:26:45.950385 {0x00000790} [info] log : logger.cpp:198 : Added logger named "conditions"
15:26:45.950385 {0x000018c4} [error] client : condition.cpp:9578 : Post-second: 015:26:45.950385 {0x00000790} [info] log : logger.cpp:331 : Setting "conditions" logger threshold to "debug".
15:26:45.950385 {0x00002310} [trace] conditions : condition.cpp:9562 : And::Eval searching non_matches with input matches (0):  and input non_matches(1): Hnub Qub II (672)  
15:26:45.950385 { } [trace] conditions : condition.cpp:9562 : And::Eval searching non_matches with input matches (0):  and input non_matches(1): Hnub Qub II (672)  15:26:45.950385 {0x00000790} [debug] log : loggerwithoptionsdb.cpp:105 : Configure log source "conditions" from optionsDB using threshold debug
15:26:45.950385 {0x00002310} [error] client : condition.cpp:9566 : Post-first: 2
15:26:45.950385 {0x00002310} [error] client : condition.cpp:9578 : Post-second: 2
15:26:45.950385 {0x000018c4} [error] client : condition.cpp:9586 : Post-third: 0
15:26:45.950385 {0x0000335c} [error] client : condition.cpp:9566 : Post-first: 0
15:26:45.950385 {0x00002310} [error] client : condition.cpp:9586 : Post-third: 2
15:26:45.950385 {0x00003ac4} [error] client : condition.cpp:9566 : Post-first: 0
15:26:45.950385 {0x00003ac4} [error] client : condition.cpp:9578 : Post-second: 0
15:26:45.950385 {0x0000335c} [error] client : condition.cpp:9578 : Post-second: 0
15:26:45.950385 {0x00000b7c} [error] client : condition.cpp:9566 : Post-first: 2
15:26:45.950385 {0x00000b7c} [error] client : condition.cpp:9578 : Post-second: 2
```
Thread ID 0xb7c and 0x2310 execute the first tracelogger statement and result in a somewhat mangled log. It only happens at the beginning of a session.

The resulting low number of calls where the logging is actually executed does not explain CPU times measured. Probably, the check for log level in a multithreaded boost logger already requires locking...

**Comment by geoffthemedio** [comment]
Does the profiling specifically mention the logger calls? It's not clear from the screenshot that the TraceLogger calls are a substantial contribution.

Or what happens if you just comment them out and run the same test to compare?

**Comment by o01eg** [comment]
@geoffthemedio  It should have lazy evaluation. I've added

```cpp
TraceLogger() << "No" << (abort(), 0);
```
which aborted application only if trace logs enabled.

**Comment by geoffthemedio** [comment]
Meaning it's working as intended and this can be closed?

Or will you try commenting out the recently-added `TraceLogger` lines to see if it makes any difference in UI responsiveness or turn processing times?

**Comment by Morlic-fo** [comment]
There were 2 issues:

1) There was the bug where trace logger statements were executed incorrectly due to temporarily ignored log level threshold. This was fixed in #2619 .

2) The logging statements still have "significant" overhead when they do the checks.

"Significant" is relative though: The only place where it shows up consistently is the And-Condition due to large the number calls (probably mostly activation clauses). This does slightly slow down turn processing time but with current number of log-statements isn't too problematic.
